### PR TITLE
[preprocessor/webhook] add browsertrix cloud webhook

### PR DESCRIPTION
**Pull request by williamchong**
Fixes https://github.com/starlinglab/integrity-v2/issues/33

TODO:
docs

### CODE REVIEW

File: webhook/browsertrix.go [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+func handleBrowsertrixEvent(w http.ResponseWriter, r *http.Request) {
+	var e BrowsertrixCrawlFinishedResponse
+	err := json.NewDecoder(r.Body).Decode(&e)
+	if err != nil {
+		writeJsonResponse(w, http.StatusBadRequest, map[string]string{"error": err.Error()})
+		return
+	}
+	if e.Event != "crawlFinished" {
+		log.Printf("Received event %s, ignoring", e.Event)
+		writeJsonResponse(w, http.StatusOK, map[string]string{})
```
Comment (makew0rld) [inline webhook/browsertrix.go:134]: Return JSON is unnecessary here, you can just use `w.WriteHeader(http.StatusOk)`.

File: webhook/browsertrix.go [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+		return
+	}
+	err = aa.SetAttestations(cid, true, attributes)
+	if err != nil {
+		log.Println("error setting attestations:", err)
+		writeJsonResponse(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
+		return
+	}
+
+	writeJsonResponse(w, http.StatusOK, map[string]string{"cid": cid})
```
Comment (makew0rld) [inline webhook/browsertrix.go:235]: Why tell Browsertrix the CID?

File: webhook/browsertrix.go [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+	FileSize      int      `json:"fileSize"`
+	FileCount     int      `json:"fileCount"`
+	Tags          []string `json:"tags"`
+	Errors        []string `json:"errors"`
+	CollectionIds []string `json:"collectionIds"`
+	// other fields are large and are omited
+}
+
+func getCrawlInfo(orgId, crawlId string) (*CrawlInfoResponse, error) {
+	jwtToken, err := getJWTToken()
```
Comment (makew0rld) [inline webhook/browsertrix.go:94]: Ideally we wouldn't be logging in each time. The returned JWT token should have an expiry date embedded in it, and `getJWTToken` could cache and re-use it until it expires.

**Review by makew0rld — CHANGES_REQUESTED** [review]
Besides docs, I see two things missing here:
- Validation of metadata: Browsertrix gives us a file size and hash, and we should check that they match our calculations.
- Authentication: the `/browsertrix` route is open to anyone. Since we can't give them a JWT to use, instead we should secure the route by having a secret in the query string. That secret can be defined in the config.

Auth example: `/browsertrix?s=wzfo63r2focmt6fw`

File: webhook/browsertrix.go [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+	if err != nil {
+		return "", err
+	}
+	var value LoginResponse
+	err = json.Unmarshal(data, &value)
+	if err != nil {
+		return "", err
+	}
+	browsertrixJwtToken = value.AccessToken
+	browsertrixJwtMutex.Unlock()
```
Comment (makew0rld) [inline webhook/browsertrix.go:86]: In all the cases where there are errors, this mutex is never unlocked. The idiomatic way of doing this in Go is to `defer` your `Unlock` right after calling `Lock`.

**Review by makew0rld — CHANGES_REQUESTED** [review]

**Comment by williamchong** [comment]
It seems browsertrix gives a new cert every time i try, not sure if we should just trust all cert sha for now

**Comment by makew0rld** [comment]
> It seems browsertrix gives a new cert every time i try, not sure if we should just trust all cert sha for now

Can you expand on this? A new `domainCert` every time?

Maybe instead of having a manual list of `trustedDomainFingerprints`, a better way to verify would be to use the OS's CA store (with `x509.SystemCertPool`), and just check that `domainCert` is a valid certificate chain for the `domain`, no matter who the CA is. Because if Browsertrix switches from Let's Encrypt to ZeroSSL one day, that shouldn't necessarily fail in our system. Is that possible? That might fix your problem.

Feel free to respond in Slack if that's better.

File: webhook/browsertrix.go [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+
+type BrowsertrixCrawlFinishedResponse struct {
+	OrgID     string `json:"orgId"`
+	ItemID    string `json:"itemId"`
+	Resources []struct {
+		Name        string `json:"name"`
+		Path        string `json:"path"`
+		Hash        string `json:"hash"`
+		Crc32       int    `json:"crc32"`
+		Size        int    `json:"size"`
```
Comment (makew0rld) [inline webhook/browsertrix.go:32]:
```suggestion
		Size        int64    `json:"size"`
```

For correctness we should explicitly support file sizes above 4 GiB.

**Review by makew0rld — APPROVED** [review]
Looks good. Happy to merge if this is tested and working on your end. Or we can do docs in this PR too.

One thing I'll add regarding docs is we should have an example curl command that sets up Browsertrix with the webhook URL.

File: webhook/browsertrix.go (hunk 5 above)
Comment (williamchong) [reply webhook/browsertrix.go]: updated and also updated comparison

**Comment by williamchong** [comment]
> Looks good. Happy to merge if this is tested and working on your end. Or we can do docs in this PR too.
> 
> One thing I'll add regarding docs is we should have an example curl command that sets up Browsertrix with the webhook URL.

Yes I have tested with actual webhook payload, will make doc in another PR

**Comment by makew0rld** [comment]
Will merge once changes mentioned in Slack are integrated.

**Comment by williamchong** [comment]
updated, we still need to have trust list for freetsa though

**Pull request merged**